### PR TITLE
Recursive & subpath based pin

### LINKED ISSUE #3477: `opam pin add -k git` in 1.2.2 vs 2.0
Testing with 2.0 rc2, a workflow involving locally pinning a package during development doesn't work in 2.0. I'm not sure if I'm currently abusing 1.2.2 or if this is unintended behavior of 2.0.

With 1.2.2, I pin packages with:
```
$ cd /path/to/git/clone/root
$ cd my-lib-subdir
$ opam pin add -n -k git my-lib .
```
and opam happily reports:
```
my-lib is now git-pinned to /path/to/git/clone/root/my-lib
[NOTE] Pinning in mixed mode: OPAM will use tracked files in the current working tree from /path/to/git/clone/root/my-lib. If this is not what you want, pin to a
       given branch (e.g. /path/to/git/clone/root/my-lib#HEAD)
```

But with 2.0 the analogous attempt fails:
```
$ opam pin add -n -k git my-lib .
[ERROR] Could not synchronize /Users/jjb/.opam2/4.07.0/.opam-switch/sources/my-lib from "git+file:///path/to/git/clone/root/my-lib#master":
        "git fetch -q file:///path/to/git/clone/root/my-lib" exited with code 128
[ERROR] Error getting source from git+file:///path/to/git/clone/root/my-lib#master
[Exit 40]
```

It seems that 1.2.2 is able to git pin a subdirectory of the git repo, while 2.0 isn't.

I tried without `-k git`, but then `opam` copies the entire directory, which is very slow and heavy as usually there are many and large files not under source control (build and test artifacts).

I also tried `opam install ./my-lib.opam --inplace-build` which initially seems to do what I'm after, but also sets up an `rsync` pin, which hits on the next e.g. `opam upgrade`.

User error?

### PR DESCRIPTION
**Pull request by rjbou**
This PR is the continuation of #3174, started by @hnrgrgr.

It introduces recursive, subpath based, and subpath recursive pin, with options `--recursive` and `--subpath PATH`.

On recursive mode, opam goes throught subdirectories to find opam files & pin them to their subdirectories.
On subpath based mode, opam goes to the given subdirectory to find opam file & pin it to this subdrectory.

Mixed pins are not allowed, pin kind is the top directory kind. If git repos are subdirectory of a local directory, thoses packages will be pinned as path based. If a subdirectory is not versionned of a top git directory, it is ignored (as it is git-based copy).

Internally, opam doesn't copy the whole top directories (except for `hg` & `darcs`), only the content of subdirectory of each subpath based package and its full directory structure from the top directory. Also, for `git+http` backend, git repository is downloaded only once, sub pins are retrieved from pin cache.

These options are available for: `pin`, `unpin`, `remove`, `install`, `reinstall`, `upgrade`, and `lint`. `opam source` also handles it, it copies the sources of the subdirectory content only.

A pin example can be found [in this gist](https://gist.github.com/rjbou/d634da7333ac0bbe40c9dfb75c4e5c6e).

~*Note: This PR is a **WIP**, it needs to be **tested**, and at the end git tree need to be cleaned before merge.*~
Closes #3174
Closes #3477

### CODE REVIEW

**Comment by rjbou** [comment]
@jberdine

> I worry about having to sync the whole repo

On this implementation, the whole repo in synced only once, at first time, then from this local pull, only package subdirectory is copied internally. After, for update check, only subdirectories are checked/retrieved (local & git backends).

> I am a bit surprised at how tightly integrated opam and git are though (e.g. also what happens if I want to use hg? :-) ).

opam handles [`hg` and `darcs`](https://opam.ocaml.org/doc/2.0/api/opam-core/OpamUrl/) VCS also. For the moment, in this PR, you can use rec/sub pin, but not have all its optimisations (full repo is cloned & checked up).
I'm not familiar with `hg` and `darcs`, any input is welcome ;)

**Comment by jberdine** [comment]
I tried testing this a bit.
I created a pin with
```
opam pin add --no-action --working-dir --inplace-build --kind=git --subpath my_package_subdir my_package ..
```
where `..` is the git repo root, `./_opam` is the dir containing the current local switch, and `../my_package_subdir` is the current dir, and `../my_package_subdir/my_package.opam` is the opam file for `my_package`.
Should I expect a copy of pretty much the full git repo's `.git` dir to be placed in `_opam/.opam-switch/sources`? I was really hoping to avoid that. Just rsyncing the files listed by git ls-files was really slick.

Also, when then trying to install the pinned package, the path the build starts from seems not to use the subpath, causing build failure. Is that expected? For example, in:
```
#=== ERROR while compiling my_package.~dev =========================================#
# context     2.0.0 | macos/x86_64 |  | pinned(git+file:///Users/jjb/root#wip#18285334)
# path        ~/root
# command     ~/.opam/opam-init/hooks/sandbox.sh build make dunes
# exit-code   2
# env-file    ~/.opam/log/my_package-7844-5f1023.env
# output-file ~/.opam/log/my_package-7844-5f1023.out
```
I expected `# path        ~/root/my_project_subdir` instead of `# path        ~/root`.

Note that there are warnings such as
```
[ERROR] Cannot sync /Users/jjb/root/ into /Users/jjb/root/my_package_subdir/_opam/.opam-switch/sources/my_package: they
        overlap
```
at various times. It seems that having a local switch in a subdir of the git repo is also trouble.

I also do not understand `--assume-built`, as with it no files seem to be installed into `../_opam/lib`, etc. at all. I interpreted the 'built' assumption as meaning built in the working tree, but not installed into the _opam switch's tree. Perhaps this is a casualty of the other steps not quite working.

**Comment by AltGr** [comment]
> `--working-dir --inplace-build`

You're not testing the simplest case here... Which is good! Thanks ;)
Note however that `inplace-build` supersedes `working-dir`, since the latter says to copy the latest version of files into the build dir, but the former tells to not use a separate build dir at all and work from the pinning source directly.

> Should I expect a copy of pretty much the full git repo's .git dir to be placed in _opam/.opam-switch/sources?

Git doesn't offer subtree cloning like e.g. SVN does. So we need to get the full `.git/`, but should then  only checkout the required subtree (@rjbou correct me if I am wrong!)
But since you used `--working-dir`, `rsync` is used instead anyways.

> Also, when then trying to install the pinned package, the path the build starts from seems not to use the subpath, causing build failure. Is that expected?

Probably a bug related to `--inplace-build`, we'll check that, thanks!

> Note that there are warnings such as
> ```
> [ERROR] Cannot sync /Users/jjb/root/ into /Users/jjb/root/my_package_subdir/_opam/.opam-switch/sources/my_package: they overlap
> ```

that's weird, we normally explicitely exclude `_opam/` from calls to `rsync` for that precise reason. Will check.

**Comment by jberdine** [comment]
> > `--working-dir --inplace-build`
>
> Note however that `inplace-build` supersedes `working-dir`, since the 
> latter says to copy the latest version of files into the build dir, 
> but the former tells to not use a separate build dir at all and work 
> from the pinning source directly.

Ah! Ok, thanks, I will try again without `--working-dir`.

>> Should I expect a copy of pretty much the full git repo's .git dir to 
>> be placed in _opam/.opam-switch/sources?
>
> Git doesn't offer subtree cloning like e.g. SVN does. So we need to 
> get the full `.git/`, but should then  only checkout the required 
> subtree (@rjbou correct me if I am wrong!)

Ok, thanks. Have you considered whether using git worktree would be 
feasible? For large repos that could save a lot of space and IO.

**Comment by jberdine** [comment]
(Sorry, this is off on a tangent about inplace-build and not subpath)

> Note however that `inplace-build` supersedes `working-dir`, since 
> the latter says to copy the latest version of files into the build 
> dir, but the former tells to not use a separate build dir at all and 
> work from the pinning source directly.

`inplace-build` without `working-dir` avoids one of the copies, nice! 
Note though that there is a warning that seems off:

```
$ opam install llvm --inplace-build
<><> Synchronising pinned packages 
 ><><><><><><><><><><><><><><><><><><><><>  🐫
[NOTE] Ignoring uncommitted changes in /Users/jjb/sat/sledge/llvm 
(`--working-dir' not active).
```

My understanding is that with `inplace-build` the uncommitted changes 
will NOT be ignored, correct?

**Comment by rjbou** [comment]
First, thanks for testing this PR. Plus, as you have a bunch of corner cases, it is a good completnes & "stress" test :).

Now the answers...

>> Should I expect a copy of pretty much the full git repo's .git dir to be placed in _opam/.opam-switch/sources?

> Git doesn't offer subtree cloning like e.g. SVN does. So we need to get the full .git/, but should then only checkout the required subtree (@rjbou correct me if I am wrong!)

Indeed, the checkout is done from the git root, but only subdirectories are checkout (with sparse checkout enabled), e.g. if you have in your repo
```
gitroot/.git:
[...]
gitroot/foo:
f1 f2
gitroot/bar:
f3 f4
```
and you pin `bar` with a subpath, you'll have in the internal opam folder
```
gitroot/.git:
[...]
gitroot/bar:
f3 f4
```

> Have you considered whether using git worktree would be feasible? For large repos that could save a lot of space and IO.

I just had a quick look to `git worktree`, it can be useful for git local pinning more generally. One of my concerns would be to depends on a new version of git (it was introduced in git 2.5). I'll look more in it!


>> Also, when then trying to install the pinned package, the path the build starts from seems not to use the subpath, causing build failure. Is that expected?
> Probably a bug related to --inplace-build, we'll check that, thanks!

It's not from the `inplace-build` option, but the pinning path. As we can see in this [example](https://gist.github.com/rjbou/d634da7333ac0bbe40c9dfb75c4e5c6e#file-rec-pin-txt-L38) the pinning path is the one of the root directory, and a subpath is also defined. In this case, the subpath was not printed in the error log, while it should.

>>  Note that there are warnings such as
```
    [ERROR] Cannot sync /Users/jjb/root/ into /Users/jjb/root/my_package_subdir/_opam/.opam-switch/sources/my_package: they overlap
```
> that's weird, we normally explicitely exclude _opam/ from calls to rsync for that precise reason. Will check.

It's because of the subpath pinning. `_opam` directory is well excluded from the rsync, and it's not even performed. A prior check fails because of the pinning path that doesn't contain the subpath (cf. over). I'll add a fix.


> My understanding is that with `inplace-build` the uncommitted changes will NOT be ignored, correct?

It's the `working-dir`, with git backend, that includes uncommitted changes (by changing the backend to rsync). The `inplace-build` performs the build locally instead of in the opam build internal directory.


> I also do not understand --assume-built, as with it no files seem to be installed into ../_opam/lib, etc. at all. I interpreted the 'built' assumption as meaning built in the working tree, but not installed into the _opam switch's tree. Perhaps this is a casualty of the other steps not quite working.

It assumes that everything has been built in the working tree, and installs it in the opam switch tree, here the local `_opam` switch.

**Comment by rjbou** [comment]
@jberdine 

>  Is `--depth 1` or similar used to limit the history that is fetched/copied?

No, for the moment, full repo is fetched. @altgr, is there a reason for not using it ?

> Are you saying that I ought to have used different paths when setting up the pin?

No, you used correctly the command, subpath was missing in the error report. The url pin with subpath is not  the full path: url pin remains the directory given as arguments, and the subpath is used internally to work with, so for display they are dissociated (cf example).

I updated the PR to adjust subpath behavior with `working-dir`: you should'nt have anymore those errors (error report / overlap / etc.).

**Comment by dra27** [comment]
Just a note that Git worktree is annoyingly not present on CentOS/RHEL installations of Git, as it still ships Git 1.8.

**Comment by rjbou** [comment]
Final update before merge!

**Comment by rjbou** [comment]
Update: no modification in the opam-file syntax and no version bump. Instead of having the `subpath` url field, it is remplaced ftm with an `x-subpath` extra field. An `opam-version` constraint is also automatically added, and lint warning and errors.

File: src/format/opamFile.ml [hunk 1]
```diff
@@ -2664,13 +2691,44 @@ module OPAMSyntax = struct
...
+         in
          match OpamPackage.of_filename filename with
          | Some nv -> with_nv nv t
          | None -> t)
       (fun (filename, t) ->
          filename,
+         let t =
+           match t.url with
+           | Some ({ URL.subpath = Some sb ; _ } as url) ->
+             add_extension t subpath_xfield (String (pos_null, sb))
```
Comment (AltGr) [inline src/format/opamFile.ml]: This could also be made conditional depending on the `opam-version:` of the file being written. Of course the default should remain `2.0` when creating files.

File: src/state/opamFileTools.ml [hunk 2]
```diff
@@ -669,7 +669,7 @@ let lint ?check_extra_files ?(check_upstream=false) t =
             if not_corresponding = [] then None
             else
             let msg =
-              Printf.sprintf "Cheksum%s %s don't verify archive"
+              Printf.sprintf "Checksum%s %s don't verify archive"
```
Comment (AltGr) [inline src/state/opamFileTools.ml]: "doesn't/don't match the archive" ?
Or better: "The archive doesn't match checksum(s) xxx"

File: src/client/opamArg.ml [hunk 3]
```diff
@@ -1155,6 +1155,18 @@ let assume_built =
...
 
+(* Options common to all path based/related commands, e.g. (un)pin, upgrade,
+   remove, (re)install *)
+let recurse =
+  mk_flag ["recursive"]
+    "Allow recursive lookups of .opam files. Cf. $(i,--subpath) also."
+
+let subpath =
+  mk_opt ["subpath"] "PATH"
+    ".opam files are retrieved from the given subpath. It can be combined with \
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/client/opamArg.ml]: yup!

File: src/repository/opamRepository.ml [hunk 4]
```diff
@@ -178,6 +178,20 @@ let pull_from_upstream
...
+          && OpamFilename.exists_dir pin_cache_dir then
+       (log "Pin cache (git) existing for %s : %s\n"
+          (OpamUrl.to_string url) @@ OpamFilename.Dir.to_string pin_cache_dir;
+        let git_cached =
+          OpamUrl.parse ~backend:`git
+          @@ OpamFilename.Dir.to_string pin_cache_dir
+        in
+        let pull =
+          let module BR = (val url_backend git_cached: OpamRepositoryBackend.S) in
+          BR.pull_url
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/repository/opamRepository.ml]: That's the point :)

File: src/state/opamUpdate.ml [hunk 5]
```diff
@@ -510,7 +510,8 @@ let download_package_source st nv dirname =
     | None   -> Done None
     | Some u ->
       (OpamRepository.pull_tree (OpamPackage.to_string nv)
-        ~cache_dir ~cache_urls ?subpath:(OpamFile.URL.subpath u)
+        ~cache_dir ~cache_urls ?working_dir
+        ?subpath:(OpamFile.URL.subpath u)
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/state/opamUpdate.ml]: tbh, i don't remember why I wrote this propagation of working dir a2e51d3f, but at that moment I'm pretty sure that I had a very good reason :D
This function `download_package_source` is not called on the update operation itself, but on graph action fetch operation (and `opam source`).

File: src/format/opamFile.ml [hunk 6]
```diff
@@ -1913,9 +1913,11 @@ module URLSyntax = struct
            (Pp.V.string -| Pp.of_module "checksum" (module OpamHash)));
       "mirrors", Pp.ppacc with_mirrors mirrors
         (Pp.V.map_list ~depth:1 Pp.V.url);
+      (* Disable for 2.1.0
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/format/opamFile.ml]: > I believe this should still be present

If ti is present, we'll have parse error if it is present, or we'll need to update opam file version.
Here, what is disabled is only the parsing printing of the supbath as an url field, not the internal representation of the field itself, which is populated by the `x-subpath` field. Linting is already checked (5bbf2fb64ae) for `x-subpath`. The idea is to keep it with only `x-subpath` field for next releases, and when a new version of the opam file can be released, at that time update opam file & linting enabling the `subpath` field itself ans its checks

File: src/format/opamFile.ml [hunk 7]
```diff
@@ -2664,13 +2691,44 @@ module OPAMSyntax = struct
            else None
          in
          let t = { t with metadata_dir } in
+         let t =
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/format/opamFile.ml]: Not really...

File: src/format/opamFile.ml (hunk 1 above)
Comment (rjbou) [reply src/format/opamFile.ml]: Same as upper comment, here, it is the subpath field that is concerned, and added with the availability restrictions. The generated file is still 2.0 and on next opam file version we won't need this `pp` workaround

File: src/format/opamFile.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (AltGr) [reply src/format/opamFile.ml]: Ah, I see. My idea was to already have the code for handling files with `opam-version: 2.1`, just not use that by default. I believe it can make the transition smoother.
- if `opam-version < 2.1`
  * on parsing, `x-subpath` is handled, and `subpath` is rejected
  * on printing, always use `x-subpath` (anyway we need to discriminate on opam-version in this case, since the field is stored in the same way in the record)
- if `opam-version >= 2.1`
  * on parsing, both are handled, but `x-subpath` is a lint warning
  * on printing, always use `subpath`

[the remark this answers is not in the document]
Comment (dra27) [reply src/format/opamFile.ml]: One thought - if opam-repository is remaining at `2.0` and therefore `opam` files in repos probably want to remain compatible with `2.0`, should we be considering the `subpath` bit itself as part of `opam-version: "3.0"` (i.e. bumping the major version of the file format at some point in the future and dropping all collected `x-` fields in the process?). i.e. **opam 2.1** supports this, but **file format** 3.0 is what will support it?

File: src/state/opamUpdate.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (AltGr) [reply src/state/opamUpdate.ml]: Ah, right, the argument is passed to the `download_package_source` function, which is specific to the package, so I guess it's just a matter for the caller to specify it correctly!

The reason why this wasn't needed is that working_dir is assumed to be only on dev packages, and these are updated through different functions (`fetch_dev_package`, etc.), and then skipped by `download_package`. This could have been incorrect if there are cases where we reach the Fetch action processing and the dev package wasn't yet cached, did you hit such a case ?

File: src/format/opamFile.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (AltGr) [reply src/format/opamFile.ml]: yeah, we were actually discussing it with @rjbou: there is no need for the file format versions to be synchronised with the opam version (and in fact, there are only loosely, e.g. a new opam version doesn't need to upgrade the file format version). Jumping to 3.0 indeed would avoid the ambiguity so I like the idea. The name of the field becomes slightly misleading, and one or two error messages (advising an upgrade of opam to version xx) should be fixed, but that's the only downside.

As for my position vs. @rjbou's: I now understand more clearly what differs. We have two ways to update:
- hard format change, as for 1.2→2.0 : an opam 1.2 file is no longer a valid opam 2.0 file, and the repository needs to be converted. Opam can still read files in the older format, but by going through a rewrite step first
- soft format change: the new format is an extension, files in the older format are still valid files in the new format.

We just added a field here, so that can fit in the second case: we don't need a rewrite to upgrade opam-repository. We _could_ hint that with a minor format version bump;

The addition of the compatibility layer through `x-foo` doesn't actually need to change that:
- when writing, encode using the new field or the `x-foo` field depending on the file format version (and add _or remove_ the availability constraint on the opam version)
- keep the old version when creating and writing files by default, for the moment.

I don't think it can hurt to have full support for files with the new `subpath` field in the code already, as long as it's not the format used by default.

[the remark this answers is not in the document]
Comment (rjbou) [reply src/format/opamFile.ml]: Yes, we can have it from now integrated in the opam file syntax. What was disturbing me from having it handled from now, is the ambiguity that we could have with an file-format at `2.1` and the corresponding opam version `2.2` or more. Increasing directly the file-format version to 3.0 is a good compromise (we will have at some point file format at 4.0 and opam at 3.0).
The _problem_ here is that we don't have the information of the `opam-version` field when we define parsing/printing url field.
For clarity: note that for the moment (in this pr), the `subpath` and `x-subpath` fields are only generated and used by/in opam internals, it is not intended to be used or manipulated by user (the only user interaction is at an `opam pin edit`)

File: src/state/opamUpdate.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (rjbou) [reply src/state/opamUpdate.ml]: I don't find a good reason to keep it as i don't remember why I added it, and don't find one now. I ran some tests, and everything is ok without. Reverted.

File: src/format/opamFile.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (dra27) [reply src/format/opamFile.ml]: I have to begin every GitHub comment at the moment with "sorry for the lag" 🙂

I agree that's unfortunate that we end up `opam-version` the field meaning the format of the file and `opam-version` the variable meaning the version of opam, but I guess the sooner that gets properly uncoupled in people's minds, the better! You already can't use the opam-version *field* to be say, `2.0.2` in order to force the package to be only readable by a maintenance release which has a bug fix you know matters. That already belongs in the available field.

We can of course have the information for `opam-version` available in the parser and lexer in future - I already in https://github.com/ocaml/opam/pull/3350/commits/78d06a8e4c0ff8fd02f4f4d6e4ef9de0a2aef55a assumed (incorrectly) that `opam-version` should be the first field. It might be sensible either to enforce that from now, or at least have `opam lint` warn if the `opam-version` field appears too late in a file.

I think my last concern with this - is still an error, therefore, for a user to provide a file which has `opam-version: "2.1"` in it? i.e. can we formally have the notion that the `2.1` is strictly internal to opam, because there are no fields requiring it for user-provided files?

**Comment by AltGr** [comment]
Just waiting for a little bit more testing before merge ^^

**Comment by rjbou** [comment]
Updated with opam-rt tests!

File: src/state/opamFileTools.ml [hunk 8]
```diff
@@ -705,6 +705,31 @@ let t_lint ?check_extra_files ?(check_upstream=false) ?(all=false) t =
...
+       | _ -> false
+     in
+     let opam_restriction =
+       OpamFilter.fold_down_left (fun acc filter ->
+           if acc then acc
+           else
+           match filter with
+           | FOp (FIdent (_, var, _), `Eq, FString version) ->
+             OpamVariable.to_string var = "opam-version" && version = "2.1"
+           | _ -> false) false t.available
```
Comment (AltGr) [inline src/state/opamFileTools.ml]: Technically, a higher constraint would be also OK. Maybe testing `simplify_version_formula (ands [v < 2.1, CSTR]) = None` could do the trick ?
